rpc/RPCTransaction.java

```diff
@@ -186,65 +182,42 @@ public void onNext(TransactionProto.Transaction.Res res) {
-                collectors.values().parallelStream().forEach(x -> x.add(new Response.Error((StatusRuntimeException) error)));
-                collectors.clear();
-                streamIsOpen.set(false);
                 transactionWasClosed.set(true);
+                streamIsOpen.set(false);
+                collectors.clearWithError(new Response.Error((StatusRuntimeException) error));
             }
 
             @Override
             public void onCompleted() {
-                collectors.values().parallelStream().forEach(x -> x.add(new Response.Error(TRANSACTION_CLOSED)));
-                collectors.clear();
                 streamIsOpen.set(false);
-                transactionWasClosed.set(true);
+                close();
             }
         };
     }
 
-    abstract static class Response {
-
-        abstract TransactionProto.Transaction.Res read();
-
-        static class Ok extends Response {
-            private final TransactionProto.Transaction.Res response;
+    private class ResponseCollectors {
+        private final ConcurrentMap<UUID, ResponseCollector> map;
 
-            Ok(TransactionProto.Transaction.Res response) {
-                this.response = response;
-            }
-
-            @Override
-            TransactionProto.Transaction.Res read() {
-                return response;
-            }
-
-            @Override
-            public String toString() {
-                return className(getClass()) + "{" + response + "}";
-            }
+        private ResponseCollectors() {
+            map = new ConcurrentHashMap<>();
         }
 
-        static class Error extends Response {
-            private final GraknClientException error;
-
-            Error(StatusRuntimeException error) {
-                // TODO: parse different gRPC errors into specific GraknClientException
-                this.error = new GraknClientException(error);
-            }
+        private ResponseCollector get(UUID requestId) {
+            return map.get(requestId);
+        }
 
-            Error(ErrorMessage error) {
-                this.error = new GraknClientException(error);
-            }
+        private synchronized void put(UUID requestId, ResponseCollector collector) {
+            if (transactionWasClosed.get()) throw new GraknClientException(TRANSACTION_CLOSED);
+            map.put(requestId, collector);
+        }
 
-            @Override
-            TransactionProto.Transaction.Res read() {
-                throw error;
-            }
+        private synchronized void remove(UUID requestId) {
+            map.remove(requestId);
+        }
 
-            @Override
-            public String toString() {
-                return className(getClass()) + "{" + error + "}";
-            }
+        private synchronized void clearWithError(Response.Error errorResponse) {
+            map.values().parallelStream().forEach(x -> x.add(errorResponse));
+            map.clear();
```

Comment:
The main reason for synchronization is the possibility of the transaction becoming closed immediately after calling `if (transactionWasClosed.get())`.

`TransactionTest` was stalling intermittently because the original implementation of response collectors (just a plain old `ConcurrentHashMap`) was not being modified in a thread-safe way. It was possible that we would add a collector and, concurrently, close the transaction which initiates a "tear-down" process, namely `clearWithError`.

Given that we therefore can't allow `put` and `clearWithError` (and presumably `remove`) to run concurrently, I believe it makes sense to extract `ResponseCollectors` as an inner class, with these three basic methods marked as `synchronized`. If we didn't use an inner class, we would have to synchronize on some arbitrary object - which would be ugly.